### ISSUE TITLE
Unit king-phisher.service entered failed state.

### ISSUE THREAD

**Issue by Deensk** [securestate/king-phisher#389]

**Comment by Deensk** [comment]
Can i get help why king-phisher server is not starting after matching PostgreSQL DB password with King-Phisher server? I just upgraded to the latest version and service wont start.   
found out it was complaining about  "CRITICAL server failed to build with error: failed to execute postgresql query 'SELECT usename FROM pg_user' via su and psql" and I ALTER USER king_phisher WITH PASSWORD 'new password'; but still wont start.


[root@d1terll king-phisher]# systemctl status king-phisher.service
● king-phisher.service - King Phisher Server
   Loaded: loaded (/usr/lib/systemd/system/king-phisher.service; enabled; vendor preset: disabled)
   Active: failed (Result: resources) since Mon 2019-07-15 19:37:43 EDT; 12min ago
  Process: 5322 ExecStart=/usr/bin/python3 /opt/king-phisher/KingPhisherServer /opt/king-phisher/server_config.yml (code=exited, status=0/SUCCESS)

Jul 15 19:37:41 root@d1terll.com systemd[1]: Starting King Phisher Server...
Jul 15 19:37:42 root@d1terll.com python3[5322]: Loading .env environment variables…
Jul 15 19:37:43 root@d1terll.com runuser[5335]: pam_unix(runuser:session): session opened for user postgres by (uid=0)
Jul 15 19:37:43 root@d1terll.com su[5357]: (to postgres) root on none
Jul 15 19:37:43 root@d1terll.com python3[5322]: CRITICAL server failed to build with error: failed to execute postgresql query 'SELECT usename FROM pg_user' via su and psql
Jul 15 19:37:43 root@d1terll.com systemd[1]: PID file /var/run/king-phisher.pid not readable (yet?) after start.
Jul 15 19:37:43 root@d1terll.com systemd[1]: king-phisher.service never wrote its PID file. Failing.
Jul 15 19:37:43 root@d1terll.com systemd[1]: Failed to start King Phisher Server.
Jul 15 19:37:43 root@d1terll.com systemd[1]: Unit king-phisher.service entered failed state.
Jul 15 19:37:43 root@d1terll.com systemd[1]: king-phisher.service failed.
[root@root@d1terll king-phisher]#

**Comment by zeroSteiner** [comment]
Please completely fill out the issue template. We need all of the details that it requests to better identify the issue and provide the help that you are requesting. Please do not open a new issue but instead just edit this one and fill out all of the details.

**Comment by Deensk** [comment]
Thanks for the info zeroSteiner, I updated the ticket with more info.
Thanks

**Comment by zeroSteiner** [comment]
@wolfthefallen can you take a look at this when you have a chance?

**Comment by wolfthefallen** [comment]
Sure I should have sometime later this week to look into this.

**Comment by Deensk** [comment]
I also wondering if we need to increase any ulimit values such us open files limit?

**Comment by zeroSteiner** [comment]
If you're still getting the same error, it looks like it has something to do with the database connection. I'm basing that on the fact that it's a query which is failing.

Can you run the server with [debug output](https://github.com/securestate/king-phisher/wiki/Frequently-Asked-Questions#how-do-i-generate-verbose-logging-output) and in the foreground, then post the logs here?

The full arguments would be `sudo ./KingPhisherServer -L DEBUG -f server_config.yml`.

**Comment by Deensk** [comment]
Below is the output i got after running the debug. 
[root@d1terll king-phisher]# sudo ./KingPhisherServer -L DEBUG -f server_config.yml
DEBUG    target directory: /opt/king-phisher
INFO     checking for the pipenv environment
DEBUG    pipenv path: '/usr/bin/pipenv'
DEBUG    pipenv Pipfile: /opt/king-phisher/Pipfile
Loading .env environment variables…
DEBUG    king phisher version: 1.13.2 (rev: 63507ae4629f) python version: 3.5.6
DEBUG    plugin dependency path: /root/.local/lib/king-phisher/python3.5/site-packages
INFO     listening on 0.0.0.0:80
INFO     serving files has been enabled
INFO     initializing database connection with driver postgresql
DEBUG    using postgresql-setup to ensure that the database is initialized
DEBUG    postgresql service is already running via systemctl
CRITICAL server failed to build with error: failed to execute postgresql query 'SELECT usename FROM pg_user' via su and psql
INFO     unloading 0 plugins
[root@d1terll king-phisher]#

**Comment by zeroSteiner** [comment]
So the issue appears to be during the initialization of the database. The command `su postgres -c psql -At -c "SELECT username FROM pg_user"` is failing for some reason. Could you drop into a root shell and run that command then post the output?

I haven't seen this error before and am not quite sure what would cause that to fail. Maybe if the `postgres` user isn't available, or if they don't have the permissions to query the database? Just a guess here, I'm thinking maybe the `postgres` user isn't available or the username is different on RHEL.

**Comment by Deensk** [comment]
No Luck 
[root@d1terll ~]# su postgres -c psql -At -c "SELECT username FROM pg_user"
su: invalid option -- 'A'
postgres service is running and i can log into su -postgress and then run psql to see all the database users.

**Comment by zeroSteiner** [comment]
You might need to do `su postgres -c 'psql -At -c "SELECT usename FROM pg_user"'` instead. Notice the single ticks around the `psql` sub command. It looks like the way I originally wrote it caused the `psql` arguments to be incorrectly interpreted as `su` arguments. I also misspelled `usename` as `username`.

For reals though, this command should give us a meaningful error: `su postgres -c 'psql -At -c "SELECT usename FROM pg_user"'`

**Comment by Deensk** [comment]
sure, make sense. So when I ran the command it asked me for password and I enter the postgres PW and got the result below 

[root@d1terll ~]# su postgres -c 'psql -At -c "SELECT usename FROM pg_user"'
could not change directory to "/root"
Password:
kpost
test1user
postgres
koopal
king_phisher
[root@d1terll ~]#

**Comment by zeroSteiner** [comment]
It prompted you for a password? That's probably the issue. I don't know why it would do that since you're running as root.

A quick google search shows that maybe if your `/etc/pam.d/su` file is missing the line `auth		sufficient	pam_rootok.so` that you might be prompted for the password even while running as root.

**Comment by Deensk** [comment]
Right, i got the prompt running it again. I will have to into my pam file to try and change that.

**Comment by Deensk** [comment]
The PAM file seems to be there already but still same asking for a password. No luck.

[root@d1terll ~]# grep pam_root /etc/pam.d/su
auth            sufficient      pam_rootok.so

**Comment by zeroSteiner** [comment]
The issue is definitely that your system is, for some reason prompting you for a password when you use `su` from the root account. I have no idea why that would be but it would definitely cause issues with the King Phisher startup routine.

Since the service is already running, I'm not sure we could get around this startup check without a code change to effectively bypass it.

**Comment by wolfthefallen** [comment]
@Deensk check your pam su file. On most linux based systems this is found at `/etc/pam.d/su`
make sure you have
`auth       sufficient pam_rootok.so`
in the file and not commented out. 

This line in the pam su configuration file allows root to `su` to another user with out being prompted for a password to that users account.

**Comment by zeroSteiner** [comment]
@wolfthefallen I think Deensk already said that was the case in the commend [here](https://github.com/securestate/king-phisher/issues/389#issuecomment-516987741).

**Comment by Deensk** [comment]
Yes, the auth sufficient pam_rootok.so seems to be ok.
Even recreating the king-phisher database still result in the same error.  
python3[31668]: CRITICAL server failed to build with error: failed to execute postgresql query 'SELECT usename FROM pg_user' via su and psql
Do I need to modify anything else in PAM or Postgres file to bypass the su command?

**Comment by zeroSteiner** [comment]
It's not going to have anything to do with the database. The issue is entirely with the fact that su is prompting your root user for a password which since it's not that PAM configuration, I have no idea why that is.

**Comment by Deensk** [comment]
Doing more research on this and still haven't found anything yet.

**Comment by True-Demon** [comment]
I'm curious if you could get around this by running `pkexec` command during the configuration steps in order to get around the sudo password prompt. This would cause the prompt to display via the local graphical interface to the user when needed. If sudo privileges are required either way, better to use a method that is guaranteed to be visible to the user.

**Comment by zeroSteiner** [comment]
Yeah but most of the time the server is installed on a headless system with no GUI.

**Comment by Deensk** [comment]
Yes, the install is in headless system with no GUI. 
I want to try reinstalling everything and see.

**Comment by Deensk** [comment]
I tried to reinstall but getting the following error below. I also try clearing command   "pipenv lock --clear" but still showing the same error.

[root@d1terll king-phisher]# sudo ./KingPhisherServer --verify-config server_config.yml
WARNING  no pre-existing pipenv environment was found, installing it now
pipenv encountered the following error:
Installing dependencies from Pipfile.lock (d316a9)…
Installing initially failed dependencies…
ERROR    failed to install the pipenv environment

**Comment by wolfthefallen** [comment]
navigate to your King-phisher folder and run:
`export PIPENV_VENV_IN_PROJECT=True && pipenv --site-packages install`
if you are having permission issues use:
`export PIPENV_VENV_IN_PROJECT=True && sudo -E pipenv --site-packages install`

**Comment by Deensk** [comment]
when i try "export PIPENV_VENV_IN_PROJECT=True && sudo -E pipenv --site-packages install"
it failed couple of times before exporting pipenv. server settings seems to passed but failed to start postgres service even though i have postgres 9.6 up and running.

[root@d1terll king-phisher]# sudo ./KingPhisherServer --verify-config server_config.yml
Loading .env environment variables…
[+] configuration verification passed
[+] all required settings are present
[root@d1terll king-phisher]#  sudo ./KingPhisherServer -L DEBUG server_config.yml
DEBUG    target directory: /opt/king-phisher
INFO     checking for the pipenv environment
DEBUG    pipenv path: '/usr/bin/pipenv'
DEBUG    pipenv Pipfile: /opt/king-phisher/Pipfile
Loading .env environment variables…
DEBUG    king phisher version: 1.14.0 (rev: 4e8de5615511) python version: 3.5.6
DEBUG    using plugin-specific library path: /var/king-phisher/lib/python3.5/site-packages
INFO     listening on 0.0.0.0:80
INFO     serving files has been enabled
INFO     initializing database connection with driver postgresql
DEBUG    using postgresql-setup to ensure that the database is initialized
DEBUG    starting external process: /usr/bin/postgresql-setup --initdb
DEBUG    starting external process: /usr/bin/systemctl status postgresql.service
INFO     postgresql service is not running, starting it now via systemctl
DEBUG    starting external process: /usr/bin/systemctl start postgresql
[root@d1terll king-phisher]# ERROR    failed to start the postgresql service via systemctl
CRITICAL server failed to build with error: postgresql service failed to start via systemctl
INFO     unloading 0 plugins

**Comment by wolfthefallen** [comment]
From the debug log you provided King Phisher is not seeing the postgresql service running, and fails to start it through systemctl.

Please trouble shoot your postgresql instance and make sure it is running through `systemctl` you can view postgresql logs with `journalctl -u postgresql.service` to view the errors of starting the service.

**Comment by Deensk** [comment]
Sure, got further on su -PostgreSQL command on startup.  "su postgres -c psql -At -c "SELECT usename FROM pg_user"

[root@d1terll king-phisher]# sudo ./KingPhisherServer -L DEBUG server_config.yml
DEBUG    target directory: /opt/king-phisher
INFO     checking for the pipenv environment
DEBUG    pipenv path: '/usr/bin/pipenv'
DEBUG    pipenv Pipfile: /opt/king-phisher/Pipfile
Loading .env environment variables…
DEBUG    king phisher version: 1.14.0 (rev: 4e8de5615511) python version: 3.5.6
DEBUG    using plugin-specific library path: /var/king-phisher/lib/python3.5/site-packages
INFO     listening on 0.0.0.0:80
INFO     serving files has been enabled
INFO     initializing database connection with driver postgresql
DEBUG    using postgresql-setup to ensure that the database is initialized
DEBUG    starting external process: /usr/bin/postgresql-setup --initdb
DEBUG    starting external process: /usr/bin/systemctl status postgresql.service
DEBUG    postgresql service is already running via systemctl
DEBUG    starting external process: su postgres -c psql -At -c "SELECT usename FROM pg_user"

**Comment by zeroSteiner** [comment]
It looks like you're running into the same issue here where it's hanging because `su` is probably prompting for a password.

**Comment by Deensk** [comment]
right, definitely don't  know why.
